Use the new_declare_or_get_parameter API for nav2_collision_monitor.

leander-dsouza opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5299  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | N/A |
| Does this PR contain AI-generated software? | No |
| Was this PR description generated by AI software? | N/A |

---

## Description of contribution in a few bullet points

* Migrated the `get_parameter` and `declare_parameter` fields to use the new `declare_or_get_parameter` API for nav2_collision_monitor.
* There are a bunch of `declare_parameter_if_not_declared` functions still left that are **not ported**, as these are declared with **no default value**. In addition, the codebase is created to raise exceptions when the default value is not assigned to these parameters.

   For instance:
   
   ```cpp
   // Leave it not initialized: the will cause an error if it will not set
   nav2::declare_parameter_if_not_declared(
     node, polygon_name_ + ".radius", rclcpp::PARAMETER_DOUBLE);
   radius_ = node->get_parameter(polygon_name_ + ".radius").as_double();
   ```
* Removed redundant `set_parameter()` calls when preceeded by `declare_parameter`.


## Description of documentation updates required from your changes

* N/A

## Description of how this change was tested

* Performed linting validation using `pre-commit run --all` and testing locally using `colcon test`

---

## Future work that may be required in bullet points

* N/A

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

leander-dsouza commented on nav2_collision_monitor/src/velocity_polygon.cpp:
```diff
@@ -73,55 +72,42 @@ bool VelocityPolygon::getParameters(
 
       // linear_min param
-      double linear_min;
       nav2::declare_parameter_if_not_declared(
```
I have not ported over any `declare_parameter_if_not_declared` functions that are declared with **no default value**. This is because the codebase is designed to [raise errors](https://github.com/ros-navigation/navigation2/pull/5588/files#diff-374ffb5feb7a7784dc5493ea6e22939a3dc70d803e4d2cc63db33c6aac7815c8R120) when they have no default value attached to them.

Moreover, I do not think it is possible to use `declare_parameter_if_not_declared` without mentioning a default value.

SteveMacenski replied:
We also have a declare and then a "get" under it, so it feels like the same design pattern. We have a version that does a type without value that you can use https://github.com/ros-navigation/navigation2/blob/main/nav2_ros_common/include/nav2_ros_common/node_utils.hpp#L203-L217

leander-dsouza replied:
Thank you for the tip, Steve.
I have migrated all instances of `declare_parameter_if_not_declared`, and migrated the `rclcpp::exceptions::ParameterUninitializedException` to `rclcpp::exceptions::InvalidParameterValueException` to match the exception raised in `declare_or_get_parameter` if the [default value](https://github.com/ros-navigation/navigation2/blob/main/nav2_ros_common/include/nav2_ros_common/node_utils.hpp#L215C31-L215C61) is not set.

SteveMacenski commented on nav2_collision_monitor/src/circle.cpp:
```diff
@@ -95,12 +95,11 @@ bool Circle::getParameters(
   try {
     // Leave it not initialized: the will cause an error if it will not set
-    nav2::declare_parameter_if_not_declared(
+    radius_ = nav2::declare_or_get_parameter<double>(
```
DOes this require the template to be explicitly included? It should be able to deduce that from the `rclcpp::PARAMETER_DOUBLE`, no?

leander-dsouza replied:
Yes, it has to be included as the template parameter cannot be deduced from `rclcpp::PARAMETER_DOUBLE`.
This is because the current definition of `declare_or_get_parameter` prevents any compile-time mapping from this enum value to the `double` type.

Probably, we can redefine the function to define types based on the parameter given.

SteveMacenski reviewed (commented):
Also, can use we a node->declare_or_get_parameter()` API instead of `nav2::declare_or_get_parameter(node, ...)`? If that's not in the node wrapped capabilities, it might be good to add that.

Can you also expand on the test changes?

leander-dsouza commented:
> Also, can use we a node->declare_or_get_parameter()`API instead of`nav2::declare_or_get_parameter(node, ...)`? If that's not in the node wrapped capabilities, it might be good to add that.
> 
I have added the missing definition of `declare_or_get_parameter` without the default value to the lifecycle node.

> Can you also expand on the test changes?

I have removed all `set_parameter()` calls that are preceded by calls to `declare_parameter()`, as these are redundant and are already set in `declare_parameter()`. In essence, the parameter is already set in the `declare_parameter()` call.
